### ISSUE TITLE
[Bug] Tensor.log() 穩定化修正使梯度與前向計算值不對應

### ISSUE THREAD

**Issue by s9213712** [s9213712/minicnn#33]
## 問題描述

`src/minicnn/nn/tensor.py` 中 `log()` 的前向和反向計算使用相同的 epsilon 偏移：

```python
def log(self) -> 'Tensor':
    out = Tensor(np.log(self.data + 1e-10), ...)  # 計算 log(x + ε)

    def _backward() -> None:
        if out.grad is not None:
            self._add_grad(out.grad / (self.data + 1e-10))  # 梯度是 1/(x + ε)
```

這裡的梯度 `1/(x + ε)` 是 `log(x + ε)` 對 `x` 的正確導數，**數學上是正確的**。

但問題在於：當用戶計算 `cross_entropy` 等損失時，若手動對 `softmax` 輸出再取 `log()`，epsilon 的存在會使梯度在 `x ≈ 0` 時比真正的 `-log(x)` 損失梯度小很多（`1/ε` vs. `1/x` 的差異），導致梯度低估。

另外，這個實作與 `log_softmax()` 的梯度計算路徑不一致，兩者組合使用時可能產生意外結果。

## 建議修復

1. 改用 `np.log1p()` 風格的數值穩定版本並加文件說明
2. 或在文件中明確說明 `log()` 內含 epsilon 偏移，提醒使用者避免 `log(softmax(x))` 的組合（應改用 `log_softmax()`）

## 相關位置

- `src/minicnn/nn/tensor.py:204-211` — `log()` 實作
- `src/minicnn/nn/tensor.py:213-237` — `log_softmax()` 使用不同路徑
